contracts/wrapper/NameWrapper.sol

```diff
@@ -511,27 +511,103 @@ contract NameWrapper is
     )
         public
-        onlyTokenOwner(parentNode)
-        canCallSetSubnodeOwner(parentNode, keccak256(bytes(label)))
         returns (bytes32 node)
     {
+
+        // Make the labelhash and node.
         bytes32 labelhash = keccak256(bytes(label));
         node = _makeNode(parentNode, labelhash);
+
+        // Make sure that IS_DOT_ETH is not part of the fuses. 
         _checkFusesAreSettable(node, fuses);
-        bytes memory name = _saveLabel(parentNode, node, label);
-        expiry = _checkParentFusesAndExpiry(parentNode, node, fuses, expiry);
 
-        if (ownerOf(uint256(node)) == address(0)) {
+        // Get the node data and parentNode data.
+        // Notice: in order to save stack space we call the parent owner baseOwner.
+        // We will use baseOwner for another purpose later. 
+        (address nodeOwner, uint32 nodeFuses, uint64 oldExpiry) = getData(uint256(node));
+        (address baseOwner, uint32 parentFuses, uint64 parentExpiry) = getData(uint256(parentNode));
+
+        // If parent controlled fuses in the node are being set, make sure CANNOT_UNWRAP
+        // is burned in the fuses of the parent node. 
+        if (fuses & PARENT_CONTROLLED_FUSES != 0 && parentFuses & CANNOT_UNWRAP == 0) {
+            revert OperationProhibited(node);
+        }
+
+        // Set the expiry between the old expiry and the parentExpiry.
+        expiry = _normaliseExpiry(expiry, oldExpiry, parentExpiry);
+
+        // Make sure the caller is the parent owner or approved, and the parent name is not expired. 
+        if ((baseOwner != msg.sender && !isApprovedForAll(baseOwner, msg.sender)) || 
+            (parentFuses & IS_DOT_ETH != 0 && parentExpiry - GRACE_PERIOD < block.timestamp)
+            ) {
+            revert Unauthorised(parentNode, msg.sender);
+        }
+
+        // We are no longer using the parent owner "baseOwner" so we now
+        // can set it to the registry owner to save stack space. 
+        baseOwner = ens.owner(node);
+
+        // If the name does not exist in the registery, check to see if it can be 
+        // created. If it does exist, then check to see if it has been wrapped.   
+        if (baseOwner == address(0)) {
+
+            // Check to make sure CANNOT_CREATE_SUBDOMAIN is not set.
+            if (parentFuses & CANNOT_CREATE_SUBDOMAIN != 0) {
+                revert OperationProhibited(node);
+            }
+
+            // If we pass all the checks, register the subname and wrap it. 
+            bytes memory _name = _saveLabel(parentNode, node, label);
             ens.setSubnodeOwner(parentNode, labelhash, address(this));
-            _wrap(node, name, owner, fuses, expiry);
+
+            // If a owner controlled fuse are being burned, check to make sure 
+            // PARENT_CANNOT_CONTROL and CANNOT_UNWRAP have been burned. 
+            _canFusesBeBurned(node, fuses);
+
+            super._mintWithData(node, owner, fuses, expiry, nodeOwner, nodeFuses, oldExpiry);
```

Comment:
This brings the gas use down to:
min max avarage
60391 · 131996 · 123610
54533 · 124304 · 116263

(-5,858) (-7,692) (-7,347)